lib/ansible/utils/ssh_agent.py

```diff
@@ -0,0 +1,666 @@
+# Copyright: Contributors to the Ansible project
+# BSD 3 Clause License (see licenses/BSD-3-Clause.txt or https://opensource.org/license/bsd-3-clause/)
+
+from __future__ import annotations
+
+import binascii
+import copy
+import dataclasses
+import enum
+import hashlib
+import socket
+import typing as t
+
+try:
+    from cryptography.hazmat.primitives import serialization
+    from cryptography.hazmat.primitives.asymmetric.dsa import (
+        DSAParameterNumbers,
+        DSAPrivateKey,
+        DSAPublicKey,
+        DSAPublicNumbers,
+    )
+    from cryptography.hazmat.primitives.asymmetric.ec import (
+        EllipticCurve,
+        EllipticCurvePrivateKey,
+        EllipticCurvePublicKey,
+        SECP256R1,
+        SECP384R1,
+        SECP521R1,
+    )
+    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
+        Ed25519PrivateKey,
+        Ed25519PublicKey,
+    )
+    from cryptography.hazmat.primitives.asymmetric.rsa import (
+        RSAPrivateKey,
+        RSAPublicKey,
+        RSAPublicNumbers,
+    )
+    from cryptography.hazmat.primitives.serialization import ssh
+
+    CryptoPublicKey = t.Union[
+        DSAPublicKey,
+        EllipticCurvePublicKey,
+        Ed25519PublicKey,
+        RSAPublicKey,
+    ]
+
+    CryptoPrivateKey = t.Union[
+        DSAPrivateKey,
+        EllipticCurvePrivateKey,
+        Ed25519PrivateKey,
+        RSAPrivateKey,
+    ]
```

Comment:
Can't we just use what's already defined in `cryptography`?
```suggestion
    CryptoPublicKey = ssh.SSHPublicKeyTypes
    CryptoPrivateKey = ssh.SSHPrivateKeyTypes
```

If we still support `cryptography < 3.4.4`, we can install https://pypi.org/p/types-cryptography into the MyPy env.

Reply:
@sivel Do you remember the reason to construct the union instead of importing the type from `ssh`?

Reply:
3.4.4 is pretty ancient- we have no plans to backport this AFAICT, so maybe a lower bound and importing would be fine?

Reply:
As @webknjaz stated, it wasn't available before 3.4.4. I'd be fine with adding a constraint to our `requirements.txt` even.

Reply:
Where does `3.4.4` come from? `ssh.SSHPublicKeyTypes` and `ssh.SSHPublicKeyTypes` were added in `40.0.0` as a [documented public API](https://github.com/pyca/cryptography/commit/79e38a379e8ab02fae6c1142db088889a0508a8c). Prior to `40.0.0` those union types were "available" as `_SSH_PUBLIC_KEY_TYPES` and `_SSH_PRIVATE_KEY_TYPES` since [`3.4`](https://github.com/pyca/cryptography/commit/f5940f068de1c07641cd102826e476531574d435). I suppose using those internal unions from old/stale versions as a fallback would be likely "fine" but 😬 .

As a data point looking at RHEL 9.4 `python3.11-cryptography` is at `37.0.2`.

One thought I had is that having our own union types shows what keys we support here, in the (however (un)likely) event `ssh.SSHPublicKeyTypes`/`ssh.SSHPrivateKeyTypes` are extended?

Reply:
@mkrizek this PyPI project README mentions 3.4.4: https://pypi.org/p/types-cryptography.

Anyway, I think it's fine if our CI is green. This would only affect type checking. And since we have `from __future__ import annotations` in the boilerplate, runtime doesn't evaluate them automatically anymore. Unless somebody sticks a library that does this, nobody's going to be impacted.